src/Symfony/Component/DependencyInjection/ServiceSubscriberTrait.php

```diff
@@ -0,0 +1,61 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection;
+
+use Psr\Container\ContainerInterface;
+
+/**
+ * @author Kevin Bond <kevinbond@gmail.com>
+ */
+trait ServiceSubscriberTrait
+{
+    /** @var ContainerInterface */
+    private $container;
+
+    public static function getSubscribedServices(): array
+    {
+        $services = array();
+
+        foreach ((new \ReflectionClass(static::class))->getMethods() as $method) {
+            if (false === strpos($method->getDocComment(), '@service')) {
+                continue;
+            }
+
+            if (!$returnType = $method->getReturnType()) {
+                continue;
+            }
+
+            $returnType = $returnType->getName();
+
+            if (!class_exists($returnType) && !interface_exists($returnType, false)) {
+                continue;
+            }
+
+            $services[$method->getName()] = $returnType;
+        }
+
+        return $services;
+    }
+
+    /**
+     * @required
+     */
+    public function setContainer(ContainerInterface $container): void
+    {
+        $this->container = $container;
+    }
+
+    protected function service(string $id)
```

Comment:
OH, for this to work setContainer needs to call the parent when available!

Reply:
> I'm not a fan of requiring subclasses to have the trait but no big deal.

there is another way to reason about this: I would really not expect the return value of this method iy vary when I extend the class where it is implemented. That'd be totally unexpected and unpreceded. Basically, I shouldn't have to care about the implementation being a static array or any other way (this trait or something else.) This is a core property to me.

> __METHOD__ includes the trait name not the class name

Wow you're that's not nice. Then we should use `$this->service(__CLASS__, __FUNCTION__);`, isn't it?

Reply:
> Then we should use $this->service(__CLASS__, __FUNCTION__);, isn't it?

OR, we could make the trait work only with public and protected methods. Any preference? Does it make sense to use a private method for these anyway?

Reply:
using function works when used in public/protected methods, but might break when mixed with private methods: a private method with same name but different return type can be defined in a parent class. When the parent will call its private method, it will get the service that was injected for the child definition. And this can just break...

Reply:
Understood. 

Using `$this->service(__CLASS__, __FUNCTION__);` would get my vote then.

Reply:
Cool. I updated my proposal above, it should be ok so you can borrow it. For the changelog, this should be in the 4.2 section. With a few test, I think we might be good :)